iwyu_include_picker.cc

```diff
@@ -1231,28 +1234,43 @@ bool MappedInclude::HasAbsoluteQuotedInclude() const {
   return IsAbsolutePath(path);
 }
 
-IncludePicker::IncludePicker(bool no_default_mappings,
-                             RegexDialect regex_dialect)
+IncludePicker::IncludePicker(RegexDialect regex_dialect,
+                             CStdLib cstdlib,
+                             CXXStdLib cxxstdlib)
     : has_called_finalize_added_include_lines_(false),
       regex_dialect(regex_dialect) {
-  if (!no_default_mappings) {
-    AddDefaultMappings();
-  }
+  AddDefaultMappings(cstdlib, cxxstdlib);
 }
 
-void IncludePicker::AddDefaultMappings() {
-  AddSymbolMappings(libc_symbol_map, IWYU_ARRAYSIZE(libc_symbol_map));
-  AddSymbolMappings(libstdcpp_symbol_map, IWYU_ARRAYSIZE(libstdcpp_symbol_map));
+void IncludePicker::AddDefaultMappings(CStdLib cstdlib,
+                                       CXXStdLib cxxstdlib) {
+  if (cstdlib == CStdLib::Glibc) {
+    AddSymbolMappings(libc_symbol_map, IWYU_ARRAYSIZE(libc_symbol_map));
+    AddIncludeMappings(libc_include_map, IWYU_ARRAYSIZE(libc_include_map));
+  }
+
+  if (cxxstdlib != CXXStdLib::None) {
```

Comment:
Does the order matter, or could we join this with the `stdlib_cpp_public_headers` block below?

Reply:
I wasn't sure about that, so I attempted to maintain the order of `AddIncludeMappings()` calls (and separately the `AddSymbolMappings()` calls). I'll do some further testing and post back.

Reply:
Yeah, good call from a pure refactoring standpoint. But if observable behavior is the same I'd love to get rid of the double take :-)

Reply:
`run_iwyu_tests.py` doesn't notice a difference - but since I'm not using `libstdcxx`, that's not much. Thinking about it, we map `from` to `to`, and order will only matter for `to` symbols for the same `from`. We're swapping the order of:

* C header -> Associated C++ header mappings and C++ private headers -> C++ public headers (and maybe a few public to public. I don't think the C headers would normally be mentioned in the latter case, so swapping these should be fine.

* Kludges for `std::size_t`, `std:allocator`, `std::char_traits` and C++ symbol -> public header mappings. The stdlib mapping can mention these types, and as long as the first mention in the canonical location (which it must be), then there shouldn't be any difference. The kludge map just means that the standard library doesn't have to worry about handling this. Current `libstdcxx` is empty, and my proposed `libc++` don't mention these types. I would expect Clang's canonical mappings to cover this.

TLDR I think we can merge the blocks.

Reply:
Nice, thanks!